src/time_tracker/invoices/core.clj

```diff
@@ -0,0 +1,56 @@
+(ns time-tracker.invoices.core
+  (:require [clojure.data.csv :as csv]
+            [time-tracker.timers.core :as timers-core]))
+
+(defn empty-time-map
+  "Returns a map of {user-id {project-id 0}}
+  for all users and projects. The last entry is the number of hours.
+  There must be at least one user and one project."
+  [users projects]
+  (->> (for [user-id (keys users)
+             project-id (keys projects)]
+         {user-id {project-id 0}})
+       (apply merge-with merge)))
+
+(defn- seconds->hours
+  [seconds]
+  (double (/ seconds 3600)))
+
+(defn add-hours
+  "Adds the hours logged against the timer entry
+  to the time map."
+  [time-map {:keys [project-id app-user-id] :as timer}]
+  (update-in time-map
+             [app-user-id project-id]
+             #(+ % (seconds->hours (timers-core/elapsed-time timer)))))
+
+(defn build-time-map
+  "Returns a map of {user-id {project-id hours-logged}}.
+  There must be at least one user and one project."
+  [users projects timers]
+  (reduce #(add-hours %1 %2)
+          (empty-time-map users projects)
+          (vals timers)))
+
+(defn- round-number
+  [number]
+  (.setScale (bigdec number) 4 java.math.BigDecimal/ROUND_HALF_UP))
+
+(defn time-map->csv-rows
+  [time-map users projects timers]
+  (for [[user-id v] time-map
+        [project-id hours] v]
+    [(:name (get users user-id))
+     (:name (get projects project-id))
+     (round-number hours)]))
+
+(defn generate-csv
+  [users projects timers]
+  (let [time-map (build-time-map users projects timers)]
+    (with-out-str
+      (csv/write-csv *out*
+                     (-> (time-map->csv-rows
+                          time-map users projects timers)
+                         (conj ["Name" "Project" "Hours Logged"]))))))
+
+(require 'time-tracker.invoices.core.spec)
```

Comment:
Strictly speaking, it doesn't need to be there at all. Since `spec`s also add to a function's docstring, I thought it might make sense to define the `spec`s right after the functions are defined. But this is unnecessary and I'll remove it.